File: web-core/src/main/java/cn/labzen/web/spring/runtime/LabzenRestResponseBodyAdvice.java
package cn.labzen.web.spring.runtime;

import cn.labzen.meta.Labzens;
import cn.labzen.web.api.response.out.Response;
import cn.labzen.web.api.response.result.Result;
import cn.labzen.web.exception.RequestException;
import cn.labzen.web.meta.WebCoreConfiguration;
import cn.labzen.web.response.format.CompositeResponseFormatter;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.core.MethodParameter;
import org.springframework.http.MediaType;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.server.ServerHttpRequest;
import org.springframework.http.server.ServerHttpResponse;
import org.springframework.http.server.ServletServerHttpRequest;
import org.springframework.http.server.ServletServerHttpResponse;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice;

import javax.annotation.Nonnull;

/**
 * REST 响应体增强处理
 * <p>
 * 在 ResponseBody 写入之前拦截，对响应内容进行格式化转换。
 * <p>
 * 核心功能：
 * <ul>
 *   <li>统一响应结构：所有 Controller 返回值都会被包装成标准 Response 格式</li>
 *   <li>异常处理：捕获 RequestException 并转换为标准响应</li>
 *   <li>格式化器链：通过 CompositeResponseFormatter 处理不同类型的返回值</li>
 * </ul>
 *
 * @see CompositeResponseFormatter
 */
@RestControllerAdvice
public class LabzenRestResponseBodyAdvice implements ResponseBodyAdvice<Object>, InitializingBean {

  private final CompositeResponseFormatter responseFormatter = new CompositeResponseFormatter();
  private boolean processAllRestResponse = true;

  /**
   * 初始化配置
   * <p>
   * 从配置中读取是否强制处理所有响应。
   */
  @Override
  public void afterPropertiesSet() {
    var configuration = Labzens.configurationWith(WebCoreConfiguration.class);
    processAllRestResponse = configuration.responseFormattingForcedAll();
  }

  /**
   * 判断是否支持当前返回值类型
   * <p>
   * 当强制格式化开启时处理所有响应，否则只处理 Result 类型。
   */
  @Override
  public boolean supports(@Nonnull MethodParameter returnType,
                          @Nonnull Class<? extends HttpMessageConverter<?>> converterType) {
    return processAllRestResponse || Result.class.isAssignableFrom(returnType.getParameterType());
  }

  /**
   * 在响应体写入之前进行格式化
   * <p>
   * 核心处理：将原始返回值通过格式化器链转换为标准响应结构。
   */
  @Override
  public Object beforeBodyWrite(Object body,
                                @Nonnull MethodParameter returnType,
                                @Nonnull MediaType selectedContentType,
                                @Nonnull Class<? extends HttpMessageConverter<?>> selectedConverterType,
                                @Nonnull ServerHttpRequest request,
                                @Nonnull ServerHttpResponse response) {
    HttpServletRequest httpRequest = (request instanceof ServletServerHttpRequest servletRequest)
      ? servletRequest.getServletRequest()
      : null;
    HttpServletResponse httpResponse = (response instanceof ServletServerHttpResponse servletResponse)
      ? servletResponse.getServletResponse()
      : null;

    if (httpRequest == null || httpResponse == null) {
      return body;
    }

    return responseFormatter.format(body, httpRequest, httpResponse);
  }

  /**
   * 处理业务异常
   * <p>
   * 捕获 Controller 中抛出的 RequestException，转换为标准响应结构。
   */
  @ExceptionHandler(RequestException.class)
  public Object handleLabzenRequestException(HttpServletRequest request,
                                             HttpServletResponse response,
                                             RequestException e) {
    return new Response(e.getCode(), e.getMessage() != null ? e.getMessage() : "internal server error", null, null);
  }
}

File: src/main/kotlin/cn/labzen/web/spring/runtime/LabzenRestResponseBodyAdvice.kt
package cn.labzen.web.spring.runtime

import cn.labzen.meta.Labzens
import cn.labzen.web.exception.RequestException
import cn.labzen.web.meta.WebConfiguration
import cn.labzen.web.response.bean.Response
import cn.labzen.web.response.bean.Result
import cn.labzen.web.response.format.CompositeResponseFormatter
import org.springframework.beans.factory.InitializingBean
import org.springframework.core.MethodParameter
import org.springframework.http.MediaType
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.server.ServerHttpRequest
import org.springframework.http.server.ServerHttpResponse
import org.springframework.http.server.ServletServerHttpRequest
import org.springframework.http.server.ServletServerHttpResponse
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * 对 Controller 返回的响应结果进行增强处理（转换 Http Response 结构）
 */
@RestControllerAdvice
class LabzenRestResponseBodyAdvice : ResponseBodyAdvice<Any>, InitializingBean {

  private val responseFormatter = CompositeResponseFormatter()
  private var processAllRestResponse = true

  override fun afterPropertiesSet() {
    val configuration = Labzens.configurationWith(WebConfiguration::class.java)
    processAllRestResponse = configuration.responseFormattingForcedAll()
  }

  override fun supports(returnType: MethodParameter, converterType: Class<out HttpMessageConverter<*>>): Boolean =
    if (processAllRestResponse) true
    else returnType.parameterType == Result::class.java

  override fun beforeBodyWrite(
    body: Any?,
    returnType: MethodParameter,
    selectedContentType: MediaType,
    selectedConverterType: Class<out HttpMessageConverter<*>>,
    request: ServerHttpRequest,
    response: ServerHttpResponse
  ): Any? {
    val httpRequest = if (request is ServletServerHttpRequest) {
      request.servletRequest
    } else return body
    val httpResponse = if (response is ServletServerHttpResponse) {
      response.servletResponse
    } else return body

    return responseFormatter.format(body, httpRequest, httpResponse)
  }

  /**
   * #第1级异常拦截处理：在这里处理业务相关异常，推荐统一封装为 [RequestException] !!
   */
  @ExceptionHandler(RequestException::class)
  fun handleLabzenRequestException(
    request: HttpServletRequest,
    response: HttpServletResponse,
    e: RequestException
  ): Any {
    return Response(e.code, e.message ?: "internal server error")
  }
}